sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala

```diff
@@ -108,6 +108,7 @@ class QueryExecution(val sparkSession: SparkSession, val logical: LogicalPlan) {
     EnsureRequirements(sparkSession.sessionState.conf),
     CollapseCodegenStages(sparkSession.sessionState.conf),
     ReuseExchange(sparkSession.sessionState.conf),
+    ReuseExchangeWithCoordinator(sparkSession.sessionState.conf),
```

Comment:
here we can put a `ReuseExchange` again.

Reply:
At this point, exchanges are not registered yet. Registration of exchanges happens after all query transformation.